tiledb/api/src/datatype/arrow.rs

```diff
@@ -46,169 +53,667 @@ impl ArrowPrimitiveTypeNative for f64 {
+    type ArrowPrimitiveType = arrow::datatypes::Int64Type;
+}
+
+impl ArrowLogicalType for UInt8Type {
+    type ArrowPrimitiveType = arrow::datatypes::UInt8Type;
+}
+
+impl ArrowLogicalType for UInt16Type {
+    type ArrowPrimitiveType = arrow::datatypes::UInt16Type;
+}
+
+impl ArrowLogicalType for UInt32Type {
+    type ArrowPrimitiveType = arrow::datatypes::UInt32Type;
+}
+
+impl ArrowLogicalType for UInt64Type {
+    type ArrowPrimitiveType = arrow::datatypes::UInt64Type;
+}
+*/
+
+/// Represents tiledb (`Datatype`, `CellValNum`) compatibility for an arrow `DataType`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DatatypeToArrowResult {
+    /// There is an exact logical match for the tiledb `Datatype`.
+    /// The individual values of the respective types have the same bit width
+    /// and are meant to be interpreted the same way.
+    ///
+    /// In general, this means that:
+    /// 1. `CellValNum::Fixed(1)` maps to an arrow primitive or date/time type.
+    /// ```
+    /// use tiledb::{array::CellValNum, datatype::arrow::DatatypeToArrowResult};
+    /// assert_eq!(DatatypeToArrowResult::Exact(arrow::datatypes::DataType::UInt8),
+    ///            tiledb::datatype::arrow::to_arrow(&tiledb::Datatype::UInt8, CellValNum::single()));
+    /// ```
+    /// 2. `CellValNum::Fixed(n) if n > 1` 1 maps to an arrow fixed size list.
+    /// ```
+    /// use arrow::datatypes::DataType as Arrow;
+    /// use tiledb::{Datatype as TileDB, array::CellValNum, datatype::arrow::DatatypeToArrowResult};
+    /// let arrow = tiledb::datatype::arrow::to_arrow(&TileDB::UInt8, CellValNum::try_from(8).unwrap());
+    /// let DatatypeToArrowResult::Exact(Arrow::FixedSizeList(item, fixed_len)) = arrow else { unreachable!() };
+    /// assert_eq!(*item.data_type(), Arrow::UInt8);
+    /// assert_eq!(fixed_len, 8);
+    /// ```
+    /// 3. `CellValNum::Var` maps to an arrow `LargeList`.
+    /// ```
+    /// use arrow::datatypes::DataType as Arrow;
+    /// use tiledb::{Datatype as TileDB, array::CellValNum, datatype::arrow::DatatypeToArrowResult};
+    /// let arrow = tiledb::datatype::arrow::to_arrow(&TileDB::UInt8, CellValNum::Var);
+    /// let DatatypeToArrowResult::Exact(Arrow::LargeList(item)) = arrow else { unreachable!() };
+    /// assert_eq!(*item.data_type(), Arrow::UInt8);
+    /// ```
+    ///
+    /// There are some exceptions, such as `(Datatype::Blob, CellValNum::Var)`
+    /// mapping to `arrow::datatypes::DataType::LargeBinary`, which is always variable-length.
+    ///
+    /// ```
+    /// use tiledb::{array::CellValNum, datatype::arrow::DatatypeToArrowResult};
+    /// assert_eq!(DatatypeToArrowResult::Exact(arrow::datatypes::DataType::LargeBinary),
+    ///            tiledb::datatype::arrow::to_arrow(&tiledb::Datatype::Blob, CellValNum::Var));
+    /// ```
+    /// When the output is any kind of list, field metadata may be used to represent the exact
+    /// input datatype if the input on its own is an inexact match.
```

Comment:
Am I readings this wrong or did you forget a negation in here? It feels like "field metadata may be used to represent the exact input datatype if the input is *not* an exact match". No idea what "on its own" means there.